### PR TITLE
[NTGDI][FREETYPE] Add Sanitizer and check memory

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Investigate the cause of memory destruction.
JIRA issue: [CORE-17324](https://jira.reactos.org/browse/CORE-17324)

- Detect double free.
- Detect uninitialized variable use.
- Detect heap collapse.

## Proposed changes

- Half-implement `ExQueryPoolBlockSize`.
- Add files `sanitizer.h` and `sanitizer.c`.
- Check memory in file `freetype.c`.
- Fill memory for check in `SanitizeExAllocatePoolWithTag`.
- Fill memory for check in `SanitizeExFreePoolWithTag`.
- Replace `ExAllocatePoolWithTag` with `SanitizeExAllocatePoolWithTag` in `freetype.c`.
- Replace `ExFreePoolWithTag` with `SanitizeExFreePoolWithTag` in `freetype.c`.

### CODE REVIEW

File: win32ss/gdi/ntgdi/sanitizer.c [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+    /* Executing has the same privileges as reading */
+    return IsBadReadPtr((LPVOID)lpfn, 1);
+}
+
+/*
+ * @implemented
+ */
+BOOL
+NTAPI
+IsBadWritePtr(IN LPVOID lp,
```
Comment (HBelusca) [inline win32ss/gdi/ntgdi/sanitizer.c:88]: You should look at https://learn.microsoft.com/en-us/windows/win32/api/winbase/nf-winbase-isbadwriteptr
> Important  This function is obsolete and should not be used. Despite its name, it does not guarantee that the pointer is valid or that the memory pointed to is safe to use. For more information, see Remarks on this page.

Not sure how that can help "sanitizing" stuff...

File: win32ss/gdi/ntgdi/sanitizer.c [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+    for (pch = psz; *pch; ++pch)
+    {
+        *pch = *pch;
+    }
+    *pch = *pch;
+}
+
+VOID FASTCALL SanitizeHeapSystem(VOID)
+{
+    // FIXME
```
Comment (katahiromz) [inline win32ss/gdi/ntgdi/sanitizer.c:91]: How to implement?

File: win32ss/gdi/ntgdi/sanitizer.c [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+        return 0;
+
+    if (P == FREED_POINTER || P == UNINIT_POINTER)
+    {
+        DPRINT1("%p is bad pointer\n", P);
+        ASSERT(0);
+        return 0;
+    }
+
+    // FIXME
```
Comment (katahiromz) [inline win32ss/gdi/ntgdi/sanitizer.c:106]: How to implement?

**Comment by katahiromz** [comment]
```txt
(win32ss/gdi/ntgdi/freetype.c:882) Fonts registry is empty.

*** Fatal System Error: 0x00000001
                       (0x000000F9,0x00000000,0x0000FFFF,0x00000000)

[7h
Entered debugger on embedded INT3 at 0x0008:0x8059D018.
kdb:> bt
Eip:
<ntoskrnl.exe:19d018 (sdk/lib/rtl/i386/debug_asm.S:56 (RtlpBreakWithStatusInstruction))>
Frames:
<ntoskrnl.exe:94b05 (ntoskrnl/ke/bug.c:1066 (KeBugCheckWithTf))>
<ntoskrnl.exe:9506f (ntoskrnl/ke/bug.c:1413 (KeBugCheckEx))>
<ntoskrnl.exe:15fce6 (ntoskrnl/include/internal/i386/trap_x.h:218 (KiSystemServiceHandler))>
<ntoskrnl.exe:3e2f (:0 (KiFastCallEntry))>
<ntdll.dll:10181>
<smss.exe:76f3>
<smss.exe:7a68>
<smss.exe:c098>
<smss.exe:c54c>
kdb:> 
```

**Comment by HBelusca** [comment]
> ```
> (win32ss/gdi/ntgdi/freetype.c:882) Fonts registry is empty.
>
> *** Fatal System Error: 0x00000001
>                        (0x000000F9,0x00000000,0x0000FFFF,0x00000000)
> ```

No idea what you have done, but it's quite messed up :rofl:
https://learn.microsoft.com/en-us/windows-hardware/drivers/debugger/bug-check-0x1--apc-index-mismatch

**Comment by katahiromz** [comment]
https://learn.microsoft.com/en-us/windows-hardware/drivers/ddi/ntifs/nf-ntifs-exquerypoolblocksize

> ExQueryPoolBlockSize is obsolete and has never been documented. This routine is exported only to support a small number of legacy drivers. If you call it in Microsoft Windows 2000 and the block is in special pool, the system will crash. Do not use this routine in your driver.

**Comment by katahiromz** [comment]
```txt
(win32ss/gdi/ntgdi/sanitizer.c:38) E11AA4FC, 00000408

*** Assertion failed: FALSE
***   Source File: ../win32ss/gdi/ntgdi/sanitizer.c, line 39

Break repeatedly, break Once, Ignore, terminate Process or terminate Thread (boipt)? 
kdb:> b
Execute '.cxr F81D11FC' to dump context
[7h
Entered debugger on embedded INT3 at 0x0008:0x8059CF6A.
kdb:> bt
Eip:
<ntoskrnl.exe:19cf6a (:0 (DbgBreakPoint))>
Frames:
<win32k.sys:ea5e7 (win32ss/gdi/ntgdi/sanitizer.c:39 (SanitizeReadPtr))>
<win32k.sys:eacce (win32ss/gdi/ntgdi/sanitizer.c:176 (SanitizePoolMemory))>
<win32k.sys:ead4b (win32ss/gdi/ntgdi/sanitizer.c:204 (SanitizeBeforeExFreePool))>
<win32k.sys:eadf6 (win32ss/gdi/ntgdi/sanitizer.c:234 (SanitizeExFreePoolWithTag))>
<win32k.sys:c6b5b (win32ss/gdi/ntgdi/freetype.c:2115 (IntLoadFontsInRegistry))>
<win32k.sys:c6fc1 (win32ss/gdi/ntgdi/freetype.c:953 (InitFontSupport))>
<win32k.sys:1674c2 (win32ss/user/ntuser/main.c:1056 (DriverEntry))>
<ntoskrnl.exe:39e67 (ntoskrnl/ex/sysinfo.c:1997 (SSISystemExtendServiceTableInformation))>
<ntoskrnl.exe:3c971 (ntoskrnl/ex/sysinfo.c:3037 (NtSetSystemInformation))>
<ntoskrnl.exe:3fe5 (:0 (KiSystemCallTrampoline))>
<ntoskrnl.exe:15fc97 (ntoskrnl/ke/i386/traphdlr.c:1840 (KiSystemServiceHandler))>
<ntoskrnl.exe:3d9b (:0 (KiSystemService))>
<ntoskrnl.exe:2380 (:0 (ZwSetSystemInformation))>
<ntoskrnl.exe:3c971 (ntoskrnl/ex/sysinfo.c:3037 (NtSetSystemInformation))>
<ntoskrnl.exe:3fe5 (:0 (KiSystemCallTrampoline))>
<ntoskrnl.exe:15fc97 (ntoskrnl/ke/i386/traphdlr.c:1840 (KiSystemServiceHandler))>
<ntoskrnl.exe:3e2f (:0 (KiFastCallEntry))>
<ntdll.dll:10181>
<smss.exe:76f3>
<smss.exe:7a68>--- Press q to abort, any other key to continue ---
```

File: win32ss/gdi/ntgdi/sanitizer.c [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+    if (bNullOK && ptr == NULL)
+        return;
+
+    ASSERT(ptr != NULL);
+    ASSERT(ptr != UNINIT_POINTER);
+    ASSERT(ptr != FREED_POINTER);
+
+    _SEH2_TRY
+    {
+        ProbeForRead(ptr, cb, 1);
```
Comment (katahiromz) [inline win32ss/gdi/ntgdi/sanitizer.c:34]: Why failing at here?

**Comment by katahiromz** [comment]
The plan is changed. I close this PR.

**Pull request closed without merging**